### PR TITLE
Added a new section on testing a program with artifical data

### PR DESCRIPTION
**Pull request by durian888**
also merged two other sections dealing with simplified data

This is my pull request required to complete instructor training - it is also my first pull request on git hub so please except my apologies if things do not go entirely to plan.

### CODE REVIEW

**Comment by abostroem** [comment]
I like the piece you added on artificial data. Any other opinions on whether this is too abstract for our audience?

I think the sections you merged could use a little more work (or could stay two sections). The two pieces are testing a simplified case: one is simplified data, the other is a simplified scenario. It would be nice to put in a some language showing that these are the same procedure for different types of code.

**Comment by durian888** [comment]
Hi Azalee.

I'm glad that you like the artificial data text - as to is whether it is too abstract - probably, but I was trying to keep it about the same length  and style as the other paragraphs. This is my first SC pull request and wanted to test the waters with a something relative small change that did not rock the boat too much. If I were not so constrained (either by time or politeness) I would have been more aggressive given the importance of validation but its relative obscurity.  I think all the paragraphs related to validation could do with more elaboration so they are more concrete. It may be better if worked through an concrete example showing the different approaches one could take all within the same context. Possibly talking through the though pattern about what to look for when thinking about validate - what can go wrong and what matters.

As to the merge of the 2 sections on simplified data and simplified scenario - I was in two minds about what to do with them - each seemed a little anemic so I decided to emphasis the start simple and build up rather than to emphasis the difference of simplified data verses simplified process. It would be fine to go back to the old pattern but each section probably would need to be beefed up.

Dominic


________________________________

File: 07-defensive.md [hunk 1]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
+	While we write our program, we can create artifical data that mimics 
+	our expected real data but exaggerates the features we wish to detect. By doing
+	this, we have a way of validating that our program is actually performing the 
+	processing we want as the exaggerated features should show a strong response
+	in the programs output. We need to create a range of artifical data so that we can 
+	confirm that our program correctly responds to a range of different inputs. 
+	Additionally we should create "null data" that consist of only noise so
+	we can check that that the program does not produce false postives. 
+	Artifical data can be made by a range of methods ranging from injecting artifical 
+	signals into your experimental setup to data generated entirely in software.
```
Comment (kynan) [inline 07-defensive.md]: I suppose the latter is more common?

File: 07-defensive.md [hunk 2]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
+	If, when tested with artifical data, our program, fails to perform as expected 
+	we need first to confirm the validity of the artificial data before investigating 
+	why our program is not behaving as expected. After being well tested, a set of
+	artifical data becomes a valuable asset as it can be used to repeatedly validate a
+	program as it evolves over time. 
+
+
+2.  *Test with simplified data or conditions*
+    Before doing statistics on a complete real data set,
+    we may do analysis with reduced data sets calculating for examples
```
Comment (kynan) [inline 07-defensive.md]: for example

File: README.md [hunk 3]
```diff
@@ -1,5 +1,6 @@
-python-novice-inflammation
+#python-novice-inflammation fake_data
```
Comment (kynan) [inline README.md]: I suppose this change is unintended?

**Comment by kynan** [comment]
I like those 2 sections. Before this is good to merge please undo the changes to the README though (bonus points for amending the commit!).

**Comment by abostroem** [comment]
@kynan Are you saying we should keep Test with simplified data and Test a simple case as two sections?

**Comment by kynan** [comment]
@abostroem Sorry no, I think those paragraphs are fine where they are.

File: 07-defensive.md [hunk 4]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
-    for two identical records,
+1.  *Test with artificial data*
+	While we write our program, we can create artifical data that mimics 
+	our expected real data but exaggerates the features we wish to detect. By doing
+	this, we have a way of validating that our program is actually performing the 
+	processing we want as the exaggerated features should show a strong response
+	in the programs output. We need to create a range of artifical data so that we can 
+	confirm that our program correctly responds to a range of different inputs. 
+	Additionally we should create "null data" that consist of only noise so
+	we can check that that the program does not produce false postives.
```
Comment (QuLogic) [inline 07-defensive.md]: There are two "that" in this sentence.

File: 07-defensive.md [hunk 5]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
+1.  *Test with artificial data*
+	While we write our program, we can create artifical data that mimics 
+	our expected real data but exaggerates the features we wish to detect. By doing
+	this, we have a way of validating that our program is actually performing the 
+	processing we want as the exaggerated features should show a strong response
+	in the programs output. We need to create a range of artifical data so that we can 
+	confirm that our program correctly responds to a range of different inputs. 
+	Additionally we should create "null data" that consist of only noise so
+	we can check that that the program does not produce false postives. 
+	Artifical data can be made by a range of methods ranging from injecting artifical
```
Comment (QuLogic) [inline 07-defensive.md]: "By a range ... ranging" sounds a bit redundant. Consider "by a variety of methods" or maybe "by several methods".

File: 07-defensive.md [hunk 6]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
+	our expected real data but exaggerates the features we wish to detect. By doing
+	this, we have a way of validating that our program is actually performing the 
+	processing we want as the exaggerated features should show a strong response
+	in the programs output. We need to create a range of artifical data so that we can 
+	confirm that our program correctly responds to a range of different inputs. 
+	Additionally we should create "null data" that consist of only noise so
+	we can check that that the program does not produce false postives. 
+	Artifical data can be made by a range of methods ranging from injecting artifical 
+	signals into your experimental setup to data generated entirely in software.  
+	If, when tested with artifical data, our program, fails to perform as expected
```
Comment (QuLogic) [inline 07-defensive.md]: The comma after "program" is not necessary.

File: 07-defensive.md [hunk 7]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
     or for some other case where we can calculate the right answer by hand.
-
-2.  *Test a simplified case.*
-    If our program is supposed to simulate
-    magnetic eddies in rapidly-rotating blobs of supercooled helium,
-    our first test should be a blob of helium that isn't rotating,
-    and isn't being subjected to any external electromagnetic fields.
-    Similarly,
-    if we're looking at the effects of climate change on speciation,
+    If our program is designed to to perform simulations, during program
```
Comment (QuLogic) [inline 07-defensive.md]: There are two "to" on this line.

File: 07-defensive.md [hunk 8]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
+	processing we want as the exaggerated features should show a strong response
+	in the programs output. We need to create a range of artifical data so that we can 
+	confirm that our program correctly responds to a range of different inputs. 
+	Additionally we should create "null data" that consist of only noise so
+	we can check that that the program does not produce false postives. 
+	Artifical data can be made by a range of methods ranging from injecting artifical 
+	signals into your experimental setup to data generated entirely in software.  
+	If, when tested with artifical data, our program, fails to perform as expected 
+	we need first to confirm the validity of the artificial data before investigating 
+	why our program is not behaving as expected. After being well tested, a set of
```
Comment (QuLogic) [inline 07-defensive.md]: "After being well tested" is odd phrasing. Maybe try "After thorough testing".

File: 07-defensive.md [hunk 9]
```diff
@@ -452,20 +452,38 @@ we'd be writing up our results and moving on to the next program.
...
+	artifical data becomes a valuable asset as it can be used to repeatedly validate a
```
Comment (QuLogic) [inline 07-defensive.md]: "as it" -> "that"

File: 07-defensive.md (hunk 2 above)
Comment (QuLogic) [inline 07-defensive.md]: Changing "should" to "may" in this sentence has a subtle, but important, difference in meaning which I'm not sure is intentional.

**Comment by QuLogic** [comment]
I do not intend to point out all of the cases, but you have misspelled "artificial" as "artifical" in several places.

**Comment by tbekolay** [comment]
This PR hasn't made progress in over a year, so I'm going to close it. I believe that the material in this PR is useful, but perhaps it would be best in a separate testing lesson (see #47 and https://github.com/swcarpentry/python-novice-gapminder/issues/6).

**Pull request closed without merging**